x-pack/legacy/plugins/maps/public/layers/styles/vector/components/style_prop_editor.js

```diff
@@ -0,0 +1,105 @@
+import { getVectorStyleLabel } from './get_vector_style_label';
+import { EuiFormRow, EuiSelect } from '@elastic/eui';
+import { VectorStyle } from '../vector_style';
+import { i18n } from '@kbn/i18n';
+
+export class StylePropEditor extends Component {
+  prevStaticStyleOptions = this.props.defaultStaticStyleOptions;
+  prevDynamicStyleOptions = this.props.defaultDynamicStyleOptions;
+
+  _onTypeToggle = () => {
+    if (this.props.styleProperty.isDynamic()) {
+      // preserve current dynmaic style
+      this.prevDynamicStyleOptions = this.props.styleProperty.getOptions();
+      // toggle to static style
+      this.props.onStaticStyleChange(
+        this.props.styleProperty.getStyleName(),
+        this.prevStaticStyleOptions
+      );
+      return;
+    }
+
+    // preserve current static style
+    this.prevStaticStyleOptions = this.props.styleProperty.getOptions();
+    // toggle to dynamic style
+    this.props.onDynamicStyleChange(
+      this.props.styleProperty.getStyleName(),
+      this.prevDynamicStyleOptions
+    );
+  };
+
+  _onFieldMetaOptionsChange = fieldMetaOptions => {
+    const options = {
+      ...this.props.styleProperty.getOptions(),
+      fieldMetaOptions,
+    };
+    this.props.onDynamicStyleChange(this.props.styleProperty.getStyleName(), options);
+  };
+
+  renderStaticDynamicSelect() {
+    const options = [
+      {
+        value: VectorStyle.STYLE_TYPE.STATIC,
+        text: this.props.customStaticOptionLabel
+          ? this.props.customStaticOptionLabel
+          : i18n.translate('xpack.maps.styles.staticDynamicSelect.staticLabel', {
+              defaultMessage: 'Fixed',
+            }),
+      },
+      {
+        value: VectorStyle.STYLE_TYPE.DYNAMIC,
+        text: i18n.translate('xpack.maps.styles.staticDynamicSelect.dynamicLabel', {
+          defaultMessage: 'By value',
+        }),
+      },
+    ];
+
+    return (
+      <EuiSelect
+        options={options}
+        value={
+          this.props.styleProperty.isDynamic()
+            ? VectorStyle.STYLE_TYPE.DYNAMIC
+            : VectorStyle.STYLE_TYPE.STATIC
+        }
+        onChange={this._onTypeToggle}
+        disabled={this.props.fields.length === 0}
+        aria-label={i18n.translate('xpack.maps.styles.staticDynamicSelect.ariaLabel', {
+          defaultMessage: 'Select to style by fixed value or by data value',
+        })}
+        compressed
+      />
+    );
+  }
+
+  render() {
+    const fieldMetaOptionsPopover = this.props.styleProperty.isDynamic() ? (
+      <FieldMetaOptionsPopover
+        styleProperty={this.props.styleProperty}
+        onChange={this._onFieldMetaOptionsChange}
+      />
+    ) : null;
+
+    return (
+      <EuiFormRow
+        label={getVectorStyleLabel(this.props.styleProperty.getStyleName())}
+        display="rowCompressed"
+      >
+        <Fragment>
+          {React.cloneElement(this.props.children, {
+            staticDynamicSelect: this.renderStaticDynamicSelect(),
+          })}
```

Comment:
This is an interesting way of doing this. It's sort of a circular dependency. `StylePropEditor` requires a child-component (color, size, ... static or dynamic editors) (1), but these child-components needs to be injected some global content as well (the static-dynamic drop down so it can enclose it nicely) (2). 

For (1), it's passed along as react-dom content (because we want to seed it with some properties, and requiring this clone here), but for (2) it's passed along as a property.

I don't know if there's an immediate "cleaner" way, but the orchestration is quite complex. This might benefit from some hand-holding by adding some clarifying comments.

One alternative to consider might be to push down the dropdown-selector all the way down to the static/dynamic editors of the individual properties, at the expense of making those components a little more bloated. So instead of including with `props.staticDynamicSelect`, it'd add it with `<StaticDynamicSelect>` and be responsible of propagating up the `_onTypeToggle`. (I don't feel strongly about this change fwiw).

Reply:
I think the current implementation is the most straight forward. It gets tricky because the component has to store the state of the previous configuration so when users change between the two, their previous configurations are not lost. If you push this into the component then that state is lost because the static or dynamic component is removed from the DOM when the toggle is applied.